pkg/provisioner/ironic/ironic.go

```diff
@@ -842,6 +862,18 @@ func (p *ironicProvisioner) startProvisioning(ironicNode *nodes.Node, checksum s
 		return result, nil
 	}
 
+	// Build target RAID configuration if needed
+	logicalDisks := []nodes.LogicalDisk{}
+	if p.host.Spec.RAID != nil {
+		logicalDisks = p.buildTargetRAIDCfg(p.host.Spec.RAID)
+	}
+
+	// Build provisioner's cleansteps for hardware configuration
+	cleanSteps := p.buildProvisionCleanSteps(logicalDisks)
+	if len(cleanSteps) != 0 {
```

Comment:
Does this work right? It seems to unconditionally trigger cleaning if RAID is ever used. What causes it to determine that the cleaning is done and it can move on to  provisioning?

Reply:
It looks very wrong, but I think it's actually correct, because both TargetProvide and TargetClean end up moving the host to Active state in Ironic, so we only need to do one or the other. We don't end up calling this function again, and we don't need to.

However, there does appear to be a bug as well: this function gets called from both Manageable and DeployFailed states and on ll. 889-893 we select either TargetProvide or TargetActive, respectively, to move the host to Active because TargetProvide is not valid in the DeployFailed state. But, TargetClean is not a valid action in the DeployFailed state either, so we should check we are not in that state before calling startManualCleaning().